# Merito e Sorte


In [2]:
import tabula
import pandas as pd

In [7]:
dados = tabula.read_pdf("/home/renata/Desktop/Cursos/projetos/Meritocracia/inscritos_2020.pdf",
                        pages = "all")
len(dados)

2

In [9]:
dados[0].head()

CODIGO                                               NOME  VAGAS  \
0     100  Administração, Ciências Contábeis, Economia, E...    187   
1     101                         Administração - Piracicaba     40   
2     105                                  Arquitetura - FAU    105   
3     110                           Arquitetura - São Carlos     31   
4     115                                      Artes Cênicas     30   

   INSCRITOS    C/V  VAGAS AC  INSCRITOS AC C/V AC  VAGAS EP  INSCRITOS EP  \
0       1562   8,35        93          1165  12,53        59           312   
1        283   7,08        22           196   8,91        11            70   
2       2724  25,94        67          1924  28,72        27           599   
3        622  20,06        15           408   27,2        10           178   
4        667  22,23        16           411  25,69         8           185   

  C/V EP  VAGAS PPI  INSCRITOS PPI C/V PPI  
0   5,29         35             85    2,43  
1   6,36          7             17    2,43  
2  22,19         11            201   18,27  
3   17,8          6             36       6  
4  23,13          6             71   11,83

In [11]:
dados[1].head()

500                             Medicina 122 15794  129,46  67 11031  \
0  501.0                     Medicina - Bauru  42  5217  124,21  26  3952   
1  505.0            Medicina - Ribeirão Preto  77  6856   89,04  55  5183   
2  510.0                 Medicina Veterinária  56  2403   42,91  38  1607   
3  515.0  Medicina Veterinária - Pirassununga  42   906   21,57  24   624   
4  520.0                             Nutrição  56   854   15,25  40   600   

  164,64  33  3484  105,58  22  1279  58,14  
0    152   8   887  110,88   8   378  47,25  
1  94,24  18  1418   78,78   4   255  63,75  
2  42,29  11   615   55,91   7   181  25,86  
3     26  10   224    22,4   8    58   7,25  
4     15  12   203   16,92   4    51  12,75

Os dados são duas páginas de uma mesma tabela, quando faz a leitura da segunda página (conforme feito acima) o tabula entende que a primeira linha é o cabeçalho e isso precisa ser modificado, pois caso não seja irá perder dados. O mesmo acontece, nesse caso, para as últimas linhas, já que elas possuem os totais.

In [13]:
#Ver últimas linhas
dados[1].tail()

500                                           Medicina         122  \
47  870.0  Química (Bacharelado e Bacharelado com Atribui...  e Qua4li2d   
48  875.0                Sistemas de Informação - São Carlos          35   
49  999.0                             Treinamento E (Exatas)         300   
50    NaN                                Total das Carreiras        9217   
51    NaN                 Total das Carreiras sem Treineiros        8317   

             15794         129,46        67  11031 164,64    33   3484 105,58  \
47  de e Ma1te7r3i  is) -4 S,1ã2o  Carlos23    112   4,87    11     52   4,73   
48             342           9,77        19    217  11,42    10    104   10,4   
49            3530          11,77       180   3177  17,65    75    308   4,11   
50          129148          14,01      5964  94888  15,91  2082  26368  12,66   
51          117019          14,07      5424  83841  15,46  1857  25450   13,7   

      22  1279 58,14  
47     8     9  1,13  
48     6    21   3,5  
49    45    45     1  
50  1171  7892  6,74  
51  1036  7728  7,46

In [14]:
#Retirando as duas últimas linhas
dados[1] = dados[1].iloc[:-2]
dados[1].tail()

500                                           Medicina         122  \
45  855.0               Química - Bacharelado e Licenciatura          84   
46  865.0            Química - Licenciatura - Ribeirão Preto          28   
47  870.0  Química (Bacharelado e Bacharelado com Atribui...  e Qua4li2d   
48  875.0                Sistemas de Informação - São Carlos          35   
49  999.0                             Treinamento E (Exatas)         300   

             15794         129,46        67 11031 164,64  33  3484 105,58  22  \
45             752           8,95        50   462   9,24  20   220     11  14   
46              61           2,18        15    46   3,07   8    10   1,25   5   
47  de e Ma1te7r3i  is) -4 S,1ã2o  Carlos23   112   4,87  11    52   4,73   8   
48             342           9,77        19   217  11,42  10   104   10,4   6   
49            3530          11,77       180  3177  17,65  75   308   4,11  45   

    1279 58,14  
45    70     5  
46     5     1  
47     9  1,13  
48    21   3,5  
49    45     1

In [15]:
#Colocando o cabeçalho correto
dados_columns = dados[0].columns
linha_extra = dados[1].columns
dados[1].columns = dados_columns
dados[1].head()

CODIGO                                 NOME VAGAS INSCRITOS     C/V  \
0   501.0                     Medicina - Bauru    42      5217  124,21   
1   505.0            Medicina - Ribeirão Preto    77      6856   89,04   
2   510.0                 Medicina Veterinária    56      2403   42,91   
3   515.0  Medicina Veterinária - Pirassununga    42       906   21,57   
4   520.0                             Nutrição    56       854   15,25   

  VAGAS AC INSCRITOS AC C/V AC  VAGAS EP  INSCRITOS EP  C/V EP  VAGAS PPI  \
0       26         3952    152         8           887  110,88          8   
1       55         5183  94,24        18          1418   78,78          4   
2       38         1607  42,29        11           615   55,91          7   
3       24          624     26        10           224    22,4          8   
4       40          600     15        12           203   16,92          4   

   INSCRITOS PPI C/V PPI  
0            378   47,25  
1            255   63,75  
2            181   25,86  
3             58    7,25  
4             51   12,75

Percebe-se que a leitura dos dados da segunda página agora contém o cabeçalho correto, porém foi perdido uma linha, a linha do apêndice. Tem que colocar ela também, para que não haja nenhuma perda de informação.

In [16]:
dados.append(pd.DataFrame([linha_extra], columns=dados_columns))
dados[2].head()

CODIGO      NOME VAGAS INSCRITOS     C/V VAGAS AC INSCRITOS AC  C/V AC  \
0    500  Medicina   122     15794  129,46       67        11031  164,64   

  VAGAS EP INSCRITOS EP  C/V EP VAGAS PPI INSCRITOS PPI C/V PPI  
0       33         3484  105,58        22          1279   58,14

Agora foi adicionado uma "terceira página" para resolver o problema da retirada da linha de medicina.

In [17]:
dados = pd.concat(dados)
dados.head()

CODIGO                                               NOME VAGAS INSCRITOS  \
0    100  Administração, Ciências Contábeis, Economia, E...   187      1562   
1    101                         Administração - Piracicaba    40       283   
2    105                                  Arquitetura - FAU   105      2724   
3    110                           Arquitetura - São Carlos    31       622   
4    115                                      Artes Cênicas    30       667   

     C/V VAGAS AC INSCRITOS AC C/V AC VAGAS EP INSCRITOS EP C/V EP VAGAS PPI  \
0   8,35       93         1165  12,53       59          312   5,29        35   
1   7,08       22          196   8,91       11           70   6,36         7   
2  25,94       67         1924  28,72       27          599  22,19        11   
3  20,06       15          408   27,2       10          178   17,8         6   
4  22,23       16          411  25,69        8          185  23,13         6   

  INSCRITOS PPI C/V PPI  
0            85    2,43  
1            17    2,43  
2           201   18,27  
3            36       6  
4            71   11,83

Utilizando a função "concat" da biblioteca pandas todas as partes das tabelas do arquivo em PDF foi colocado em apenas uma tabela.

In [18]:
len(dados)

109

Antes o tamanho do conjunto "dados" era mostrado apenas o tamanho 2, que era o número de páginas. Após adicionar todas as partes da tabela em uma só o tamanho mostrado é de quantos cursos possui, que nesse caso são 109 cursos. 